### ISSUE TITLE
Update compilers for C++20 support

### ISSUE THREAD

**Issue by DanRStevens** [lairworks/nas2d-core#826]
Would be nice to start using c++20 features. This of course would require the CI builds to be using compilers with C++20 support.

A quick test with enabling C++20 results in successful builds for:
- `build-linux-gcc`  (Ubuntu 20.04)
- `build-linux-clang`  (Ubuntu 20.04)

Enabling C++20 results in build failures for:
- `build-linux-mingw`  (Ubuntu 20.04)
- `build-linux`  (Ubuntu 18.04)
- `build-macos`

> error: unrecognized command line option '-std=c++20'

I did not test updating MSVC to enable C++20 support.

----

We can drop the build for Ubuntu 18.04. That image is getting a bit old now. It's probably not reasonable to expect C++20 features on Ubuntu 18.04.

We may need to look into updating the image for Mingw. It is based on GCC, so if GCC has been updated to support C++20, there will probably be corresponding support in an updated Mingw.

On MacOS, the C++ compiler shipped with Xcode is based on Clang, though tends to lag significantly behind the latest Clang release. On CircleCI we are currently using the MacOS image with Xcode `12.3.0`, which apparently is based on Clang `12.0.0`. It seems the `c++20` option should have been added in Clang 10, so not quite sure what the problem is here.

**Comment by DanRStevens** [comment]
It's seems the `c++2a` standard (pre-c++20 official release) is available for most of the compilers. Not sure we want to target a standard pre-release though. Probably better to just wait until there is widespread compiler support for the new `c++20` standard before upgrading. That might  mean revisiting this in about a year.

**Comment by cugone** [comment]
VS 16.10+ (VS2019) fully supports C++20 when the  `/std:c++latest` language switch is supplied. Future versions (VS2022) will support a C++20 switch `/std:c++20` explicitly. See: https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-160

**Comment by ldicker83** [comment]
Was looking into VS2022, looks pretty promising. Looking forward to upgrading!

The big drag here is probably going to be macOS X support. As always Apple will probably drag their feet unnecessarily long.

**Comment by cugone** [comment]
> Was looking into VS2022, looks pretty promising. Looking forward to upgrading!
> 
> The big drag here is probably going to be macOS X support. As always Apple will probably drag their feet unnecessarily long.

Yeah they mentioned that 16.11 will be the last major update to VS2019. I can't wait for VS2022 and finally get to use the C++20 language switch and features! I don't like using the `/std:latest` switch because they explicitly state things can change. I'd rather wait for a stable explicit `/std:c++20` switch.

Game development for Mac is always a big hassle. That's why most don't do it and just tell mac users to use Wine or some other emulator. 99% of the time it works.

**Comment by DanRStevens** [comment]
Good to hear progress is being made on the Windows side.

I did an updated run for the Linux and MacOS builds in #888 using `c++20` and `c++2a`. Build links:
- [Using `c++20`](https://app.circleci.com/pipelines/github/lairworks/nas2d-core/1641/workflows/f9be409d-8cf5-4ffc-a9a9-7e829ecece5e)
- [Using `c++2a`](https://app.circleci.com/pipelines/github/lairworks/nas2d-core/1642/workflows/08c5e339-f5c1-4559-ba0c-2b2073849fe4)

It seems the most recent version of MacOS is now able to use the `--std=c++20` flag.

Not surprisingly the now somewhat outdated compiler for Ubuntu 18.04 doesn't handle any C++20 features. I'm mildly hesitant to drop that build yet, since Ubuntu 18.04 is a LTS release, which doesn't reach end of life until 2023.

There's also Mingw-w64, which is lagging behind, with no support yet for `--std=c++20`, though it is able to use `--std=c++2a`. I checked briefly if there was an updated Mingw-w64 compiler available for Ubuntu 20.04, though sadly no.

**Comment by DanRStevens** [comment]
Tagging issue #999, which relates specifically to Visual Studio 2022. The Windows build images now include Visual Studio 2022, which should have c++20 support, though I don't believe the Windows builds are configured to make use of c++20 yet.

I should mention that the `makefile` was updated awhile back to use `c++20` mode. All the Linux builds images, and the MacOS build environment had been updated to use compilers with c++20 support. The Linux builds have been further updated recently to be based on Ubuntu 22.04. The compilers from the stock `apt` packages there all have c++20 support.
